File: intermediate-package/writing-integ-tests-for-spring-web-apps/rest-api/spring-boot/src/main/java/com/testwithspring/intermediate/web/TaskCrudController.java
package com.testwithspring.intermediate.web;

import com.testwithspring.intermediate.task.TaskCrudService;
import com.testwithspring.intermediate.task.TaskDTO;
import com.testwithspring.intermediate.task.TaskFormDTO;
import com.testwithspring.intermediate.task.TaskListDTO;
import com.testwithspring.intermediate.user.LoggedInUser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

/**
 * This controller implements the REST API that provides CRUD
 * operations for tasks.
 */
@RestController
@RequestMapping("/api/task")
public class TaskCrudController {

    private static final Logger LOGGER = LoggerFactory.getLogger(TaskCrudController.class);

    private final TaskCrudService crudService;

    @Autowired
    TaskCrudController(TaskCrudService crudService) {
        this.crudService = crudService;
    }

    /**
     * Creates a new task.
     * @param task  The information of the created task.
     * @return      The information of the created task.
     */
    @RequestMapping(method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.CREATED)
    public TaskDTO create(@Valid @RequestBody TaskFormDTO task, @AuthenticationPrincipal LoggedInUser loggedInUser) {
        LOGGER.info("Creating a new task with information: {}", task);

        TaskDTO created = crudService.create(task, loggedInUser);
        LOGGER.info("Created a new task with information: {}", created);

        return created;
    }

    /**
     * Deletes an existing task.
     * @param taskId    The id of the deleted task.
     * @return          The information of the deleted task.
     * @throws com.testwithspring.intermediate.common.NotFoundException if the taks is not found.
     */
    @RequestMapping(value = "{taskId}", method = RequestMethod.DELETE)
    public TaskDTO delete(@PathVariable("taskId") Long taskId) {
        LOGGER.info("Deleting task with id: {}", taskId);

        TaskDTO deleted = crudService.delete(taskId);
        LOGGER.info("Deleted task: {}", deleted);

        return deleted;
    }

    /**
     * Finds all tasks.
     * @return  A list of tasks found from the database. If no tasks is found,
     * this method returns an empty list.
     */
    @RequestMapping(method = RequestMethod.GET)
    public List<TaskListDTO> findAll() {
        LOGGER.info("Finding all tasks");

        List<TaskListDTO> tasks = crudService.findAll();
        LOGGER.info("Found {} tasks", tasks.size());

        return tasks;
    }

    /**
     * Finds a single task.
     * @param taskId    The id of the requested task.
     * @return          The information of the found task.
     * @throws com.testwithspring.intermediate.common.NotFoundException if the taks is not found.
     */
    @RequestMapping(value = "{taskId}", method = RequestMethod.GET)
    public TaskDTO findById(@PathVariable("taskId") Long taskId) {
        LOGGER.info("Finding task with id: {}", taskId);

        TaskDTO found = crudService.findById(taskId);
        LOGGER.info("Found task: {}", found);

        return found;
    }

    /**
     * Updates the information of an existing task.
     * @param task  The new information of an existing task.
     * @return      The information of the updated task.
     * @throws com.testwithspring.intermediate.common.NotFoundException if the taks is not found.
     */
    @RequestMapping(value= "{taskId}", method = RequestMethod.PUT)
    public TaskDTO update(@PathVariable("taskId") Long taskId,
                          @Valid @RequestBody TaskFormDTO task,
                          @AuthenticationPrincipal LoggedInUser loggedInUser) {
        task.setId(taskId);
        LOGGER.info("Updates an existing task with information: {}", task);

        TaskDTO updated = crudService.update(task, loggedInUser);
        LOGGER.info("Updated the information of a task: {}", updated);

        return updated;
    }
}

File: master-package/bonus-webinars-2019/unit-tests-rest-api/src/main/kotlin/com/testwithspring/master/web/TaskCrudController.kt
package com.testwithspring.master.web

import com.testwithspring.master.task.CreateTaskDTO
import com.testwithspring.master.task.TaskCrudService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import com.testwithspring.master.task.TaskDTO
import com.testwithspring.master.task.TaskListItemDTO
import com.testwithspring.master.user.LoggedInUser
import org.springframework.http.HttpStatus
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

/**
 * Implements a REST API which provides CRUD operations
 * for tasks.
 */
@RestController
@RequestMapping("/api/task")
open class TaskCrudController(@Autowired private val service: TaskCrudService) {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(TaskCrudController::class.java)
    }

    /**
     * Creates a new task.
     * @param   input           The information of the created task.
     * @param   loggedInUser    The authenticated user.
     * @return  The information of the created task.
     */
    @RequestMapping(method = [RequestMethod.POST])
    @ResponseStatus(HttpStatus.CREATED)
    fun create(
            @RequestBody @Valid input: CreateTaskDTO,
            @AuthenticationPrincipal loggedInUser: LoggedInUser
    ): TaskDTO {
        LOGGER.info("User: #{} is creating a new task with information: {}",
                loggedInUser.id,
                input
        )

        val returned = service.create(input, loggedInUser)

        LOGGER.info("User: #{} created a new task with information: {}",
                loggedInUser.id,
                returned
        )

        return returned
    }

    /**
     * Finds the information of all tasks.
     * @return  A list that contains the found tasks. If no tasks
     *          are found, this function returns an empty list.
     */
    @RequestMapping(method = [RequestMethod.GET])
    fun findAll(): List<TaskListItemDTO> {
        LOGGER.info("Finding all tasks")

        val tasks = service.findAll()
        LOGGER.info("Found {} tasks")

        return tasks;
    }

    /**
     * Finds the information of the specified task.
     * @param id    The id of the requested task.
     * @return      The information of the found task.
     * @throws com.testwithspring.master.common.NotFoundException if no task is found with the given id
     */
    @RequestMapping(value = ["{id}"], method = [RequestMethod.GET])
    fun findById(@PathVariable("id") id: Long): TaskDTO {
        LOGGER.info("Finding task with id: {}", id)

        val found = service.findById(id)
        LOGGER.info("Found task: {}", found)

        return found
    }
}